### ISSUE TITLE
ActionController::Streaming doesn't stream

### ISSUE THREAD

**Issue by nateberkopec** [rails/rails#23828]
`ActionController::Streaming` appears to not actually stream anything on Rails master. 

[Reproduction script here](https://gist.github.com/nateberkopec/bc671b0dcc07917099c4). Run it as a rack app with `puma config.ru`. 

### What I Expect

Viewing this page in the browser should immediately show "Starting..." and then "1", "2", "3" etc should appear gradually as the bytes are streamed to the browser.

### What Actually Happens

Nothing is displayed until the entire response has completed rendering. 

This can be confirmed in Chrome's net-internals tab. The request is sent at 3507 milliseconds, and even the *headers* are not received until 9528 milliseconds.

<img width="556" alt="screen shot 2016-02-23 at 2 32 22 pm" src="https://cloud.githubusercontent.com/assets/845662/13264046/5ef9f646-da3a-11e5-831b-a08a96f78308.png">

![screen shot 2016-02-23 at 2 36 48 pm](https://cloud.githubusercontent.com/assets/845662/13264161/e587515e-da3a-11e5-97c2-36fc28d29c66.png)

I've tried a lot of different permutations of this. At first I thought maybe it was my webserver - nope, same behavior with Puma, Webrick and Unicorn (even with the fancy config or whatever you're supposed to use). Then I thought it was because Rails might be rendering the layout, then flushing, then rendering the view template, then flushing, then back to the layout before flushing again - but no combination of `sleep`s anywhere produced streaming output.

**Comment by nateberkopec** [comment]
I should note I discovered this when messing with Rubygems.org, and it doesn't work there either, so I think this is broken on 4.2 too.

**Comment by maclover7** [comment]
cc @tenderlove

**Comment by maclover7** [comment]
Reproduced on master -- this may be related to #23659.

**Comment by nateberkopec** [comment]
Doesn't appear to work on 4.2 either. [Here's a full 4.2 app that doesn't stream with Puma.](https://github.com/nateberkopec/streamer_repro/commit/842888907f56b69e1d0960192ea177f47e977d46)

I feel like I must be missing something obvious with this feature, there's no way this just "doesn't work" right?

**Comment by tenderlove** [comment]
@nateberkopec I suspect it's to do with using an inline template.  Probably just doesn't take the same code path.  I'll poke at it.

**Comment by nateberkopec** [comment]
I don't think so. Can repro with non-inline templates too, see the linked 4.2 app.

**Comment by tenderlove** [comment]
So basically what I've found is that this has never worked.  I also found an annoying thing about `stream: true`.  The only time `stream: true` does anything is [if you have a layout name](https://github.com/rails/rails/blob/2c02bc0a47777ad8cf98e1465c08b1a68151803e/actionview/lib/action_view/renderer/streaming_template_renderer.rb#L47), which is very confusing especially because there are conditionals [here](https://github.com/rails/rails/blob/2c02bc0a47777ad8cf98e1465c08b1a68151803e/actionview/lib/action_view/renderer/streaming_template_renderer.rb#L50) and [here](https://github.com/rails/rails/blob/2c02bc0a47777ad8cf98e1465c08b1a68151803e/actionview/lib/action_view/renderer/streaming_template_renderer.rb#L69) with [comments](https://github.com/rails/rails/blob/2c02bc0a47777ad8cf98e1465c08b1a68151803e/actionview/lib/action_view/renderer/streaming_template_renderer.rb#L72-L75) which make it seem like "not having a layout" is a supported feature.  The conditional requiring a layout name was added in 29078ff8.  If you remove that conditional, it still won't stream like you would expect though, and I think that's due to the template [being rendered with the default output stream](https://github.com/rails/rails/blob/2c02bc0a47777ad8cf98e1465c08b1a68151803e/actionview/lib/action_view/renderer/streaming_template_renderer.rb#L91).  It's extremely annoying to figure out what the default output buffer object is because that code is generated along with the generated method for the template.  The method that holds the compiled template can be found [here](https://github.com/rails/rails/blob/2c02bc0a47777ad8cf98e1465c08b1a68151803e/actionview/lib/action_view/template.rb#L287-L293).  Our ERB processor adds a preamble that defaults the output buffer isn't set [here](https://github.com/rails/rails/blob/2c02bc0a47777ad8cf98e1465c08b1a68151803e/actionview/lib/action_view/template/handlers/erb.rb#L9).  So every ERB template has that line at the top (I think we can fix that).

Anyway, if we remove the "layout name" conditional, and pass the streaming buffer to the render method, then it will stream for me:

```diff
diff --git a/actionview/lib/action_view/renderer/streaming_template_renderer.rb b/actionview/lib/action_view/renderer/streaming_template_renderer.rb
index f38e276..e0756ff 100644
--- a/actionview/lib/action_view/renderer/streaming_template_renderer.rb
+++ b/actionview/lib/action_view/renderer/streaming_template_renderer.rb
@@ -44,7 +44,7 @@ module ActionView
     # object that responds to each. This object is initialized with a block
     # that knows how to render the template.
     def render_template(template, layout_name = nil, locals = {}) #:nodoc:
-      return [super] unless layout_name && template.supports_streaming?
+      return [super] unless template.supports_streaming?
 
       locals ||= {}
       layout   = layout_name && find_layout(layout_name, locals.keys, [formats.first])
@@ -88,7 +88,7 @@ module ActionView
         # from the template, so start rendering it. If not, it means
         # the layout exited without requiring anything from the template.
         if fiber.alive?
-          content = template.render(view, locals, &yielder)
+          content = template.render(view, locals, output, &yielder)
 
           # Once rendering the template is done, sets its content in the :layout key.
           view.view_flow.set(:layout, content)
```

I haven't run the tests yet though.  Anyway, there's TMI on how templates get handled.  I'll try to complete the patch and fix this.

**Comment by tenderlove** [comment]
@nateberkopec I think [this line is wrong](https://github.com/nateberkopec/streamer_repro/commit/842888907f56b69e1d0960192ea177f47e977d46#diff-aa8d72ae906ee99733b90948455de8c8R3). I think it should be `stream`.

**Comment by nateberkopec** [comment]
@tenderlove corrected. now it renders the layout, stops, then renders the entire template, which I think is consistent with everything you've said above.

**Comment by nateberkopec** [comment]
Your patch works for me too. I wonder if partial streaming would be difficult to implement? StreamingTemplateRenderer could really use some love :(

**Comment by PikachuEXE** [comment]
@nateberkopec 
I guess this can be closed?

I agree that response streaming doesn't work well in some cases.
E.g. [H2 Early Hints](https://github.com/rails/rails/pull/30744)

**Comment by seanlinsley** [comment]
HTTP supports sending headers at the end of a streaming request, so there's no reason that H2 Early Hints couldn't work with template streaming.

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Trailer

I've been subscribed to this issue but haven't looked into it too much. Reading it now, it seems the issue should be renamed to clarify that streaming doesn't work when a layout isn't set.

**Comment by PikachuEXE** [comment]
Early hint headers are meant to be sent before response content, otherwise it has little value.
I guess if streaming is used the client will parse the received content and detect required resources anyway. Then it's not necessary for early hints?

**Comment by nateberkopec** [comment]
I don't think this can be closed, because the feature doesn't do what it's supposed to.

**Comment by PikachuEXE** [comment]
Oh I misread the original post
Let me try the `net-internals` tab later

**Comment by seanlinsley** [comment]
IIRC, I could only get streaming to work when I had a layout that had multiple `yield`s. It seemed to be the case that one chunk would be sent for every `provide` block in the template.

**Comment by seanlinsley** [comment]
Chrome's net-internals tab seems to back this up. It's showing 7 `URL_REQUEST_JOB_BYTES_READ`, which is the same number of `yield` & `provide` calls.

**Comment by sfcgeorge** [comment]
My findings:

Streaming of the layout works. AKA your head content (stylesheets and whatnot) will stream, then the view `<%= yield %>` always blocks. That's what the demo app here demonstrates.

It would of course be nice if views and partials streamed bit by bit too, then slow DB calls wouldn't block all the content. Nonetheless, streaming in assets and maybe your navigation first is still very useful as the browser can parse the JS and CSS early ready for the content (especially with huge modern JS frameworks that take a while to parse on mobile).

Things that break it:

* `include ActionController::Live` bafflingly you can't use Live and streaming templates in the same controller even though they seem to be 2 sides of the same coin.
* HAML, Hamlit, Faml templates / views. Erubi and Slim do support streaming. However HAML can still be used in partials without blocking streaming, just can't be used for your layout & views.
* Nested layouts as per the [Rails Guide](https://guides.rubyonrails.org/layouts_and_rendering.html#using-nested-layouts) because putting `yield` within `content_for` blocks streaming. Instead have a `application_before.html.haml` and `application_after.html.haml` to extract common layout bits into, so you don't need to nest.
* `content_for` (just switch to `provide`, easy change, _nearly_ same semantics).
* Unmet `yield :foo`s. AKA if you have a `yield :bar` at the top of your template but the `provide :bar` is within your view then it blocks. Try to change your ordering or unnest views.
* Rack::Deflate middleware. But you probably don't need it with a CDN.
* Webpacker! Doesn't just break streaming but spits garbage into the page, I have [submitted a fix](https://github.com/rails/webpacker/pull/2196).

@tenderlove's suggestion `content = template.render(view, locals, output, &yielder)` does get views and partials to output their contents early without blocking, unfortunately in the wrong order — the layout comes out after.

---

So provided Webpacker gets fixed I plan to get our app streaming the head, that'll be a really nice little performance gain site wide for almost no work 😄

I'd guess almost nobody has used streaming since 2016 which is a shame. Perhaps because it's quite invisible when working, and hard to debug when not. It seems like a really great feature though.

**Comment by jakeNiemiec** [comment]
> I'd guess almost nobody has used streaming since 2016 which is a shame. Perhaps because it's quite invisible when working

![image](https://user-images.githubusercontent.com/16782063/62168617-94803680-b2eb-11e9-9927-f9afdd5d1aa9.png)

@sfcgeorge HTTP/2 does not support chunked transfer encoding. ([ref1](https://http2.github.io/http2-spec/#rfc.section.8.1) [ref2](https://en.wikipedia.org/wiki/HTTP/2#Differences_from_HTTP_1.1)) 

> there's no reason that H2 Early Hints couldn't work with template streaming.

@PikachuEXE @seanlinsley Early Hints require HTTP/2.

**Comment by sfcgeorge** [comment]
@jakeNiemiec I saw HTTP/2 (H2) doesn't support chunking which initially was a bummer. But from some very brief reading it sounds like H1.1 to H2 proxies can convert H1.1 chunks to H2 streams... 

What I'm going to try doing is running Rails over standard H1.1 with chunking enabled, then proxy it to H2 and hopefully get the best of both worlds without rewriting anything. CloudFlare as a free H1.1 to H2 proxy option (via NGINX I believe) so I'll try that first as low barrier to entry 🎉 

If someone fancies rewriting Rails chunking to work with H2 streams (or my above idea doesn't work) that would be awesome! Unfortunately [Rack doesn't support H2 yet](https://www.speedshop.co/2016/01/07/what-http2-means-for-ruby-developers.html) so that would also need rewriting 😬

**Comment by ioquatix** [comment]
The idea that HTTP/2 doesn't support streaming because it doesn't support `transfer-encoding: chunked` is wrong. ALL HTTP/2 streams are full-duplex "chunked" which is why "transfer-encoding: chunked" must not be used because it doesn't make any sense for HTTP/2 which uses binary framing to sent "chunks".

**Comment by ioquatix** [comment]
Since the release of Rack 3 standardises the underlying streaming mechanism, and I added a convenience method `#<<` must be supported by the output interface, in most cases it should be possible to use a streaming body and use the given stream argument to the ERB (or other template engine) for full streaming support.
